### PR TITLE
fix issue 255

### PR DESCRIPTION
**Pull request by londumas**
https://github.com/igmhub/picca/issues/255

Remove delta too far from last or first object

### CODE REVIEW

**Comment by ngbusca** [comment]
@londumas could you give the performance gain from this PR?

**Comment by londumas** [comment]
I think we gain only a little with this. But it is more coherent with other codes, and a little is also nice.

**Comment by londumas** [comment]
Let me find you a situation where it is important, I want to convince you.

**Comment by londumas** [comment]
A 10% gain ! This is incredible, it deserves a reward.
More seriously, it isn't very important but it helps.

#
#MASTER
#

# angmax=0.0471522019075


[u6011908@elmo:Codes]$ time do_metal_xdmat.py --in-dir ..//DR14/Delta_LYA/Delta/ --out master.fits.gz --drq /uufs/astro.utah.edu/common/home/u6011908/Data/Catalogs/DR14Q_v3_1__THINGID_DR7.fits --lambda-abs 1215.67 --lambda-abs-name LYA --abs-igm LYA --rp-min -200.0 --rp-max +200.0 --rt-max +200.0 --np 100 --nt 50 --rej 0. --nspec 10000 --z-min-obj 4.

computing metal dmat LYA
real	0m42.860s
user	0m36.598s
sys	0m6.577s


#
# FIX ISSUE
#

# ang max = 0.0429125216136

[u6011908@elmo:Codes]$ time do_metal_xdmat.py --in-dir ..//DR14/Delta_LYA/Delta/ --out issue255.fits.gz --drq /uufs/astro.utah.edu/common/home/u6011908/Data/Catalogs/DR14Q_v3_1__THINGID_DR7.fits --lambda-abs 1215.67 --lambda-abs-name LYA --abs-igm LYA --rp-min -200.0 --rp-max +200.0 --rt-max +200.0 --np 100 --nt 50 --rej 0. --nspec 10000 --z-min-obj 4.

reading qsos
231 of 232

computing metal dmat LYA
real	0m40.215s
user	0m32.827s
sys	0m7.244s

**Comment by londumas** [comment]
Same as https://github.com/igmhub/picca/pull/66 and https://github.com/igmhub/picca/issues/44

**Comment by ngbusca** [comment]
Ok, 10% is 10%!

File: bin/do_metal_xdmat.py [hunk 1]
```diff
@@ -136,8 +136,47 @@ def calc_metal_xdmat(abs_igm,p):
...
+    ### Remove pixels if too far from objects
+    zmin_obj = None
+    zmax_obj = None
+    for pix in xcf.objs:
+        for q in xcf.objs[pix]:
+            if zmin_obj is None: zmin_obj = q.zqso
+            if zmax_obj is None: zmax_obj = q.zqso
+            zmin_obj = min(zmin_obj,q.zqso)
+            zmax_obj = max(zmax_obj,q.zqso)
+    if ( zmin_obj is not None and ( (zmin_pix<zmin_obj) or (zmax_obj<zmax_pix)) ):
```
Comment (ngbusca) [inline bin/do_metal_xdmat.py]: zmin_obj can't be None at this point. Can you simplify the expression?

File: bin/do_metal_xdmat.py [hunk 2]
```diff
@@ -136,8 +136,47 @@ def calc_metal_xdmat(abs_igm,p):
...
+                        d.ll = d.ll[w]
+                        d.co = d.co[w]
+                        d.r_comov = d.r_comov[w]
+
+    ### Define angmax
+    zmin_pix = None
+    for pix in xcf.dels:
+        for d in xcf.dels[pix]:
+            if zmin_pix is None: zmin_pix = d.z.min()
+            zmin_pix = sp.append([zmin_pix],d.z).min()
```
Comment (ngbusca) [inline bin/do_metal_xdmat.py]: can you simplify this: 

    if zmin_pix > d.z[0]:
        zmin_pix = d.z[0]

?

File: bin/do_metal_xdmat.py (hunk 1 above)
Comment (londumas) [reply bin/do_metal_xdmat.py]: Yes it can if there are no objetc in the catalog for the required zmin_pix and zmax_pix

**Pull request closed without merging**